File: miniapp/backend/routers/favorites.py
from typing import List, Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from core.responses import success_response
from crud.favorite import (
    add_favorite,
    batch_delete_favorites,
    delete_favorite,
    list_favorites,
)
from deps import get_current_user, get_db
from models import User
from schemas.favorite import Favorite as FavoriteSchema

router = APIRouter(prefix="/api/favorites", tags=["favorites"])


class FavoriteCreateRequest(BaseModel):
    protocol_name: str
    protocol_type: Optional[str] = None


class FavoriteBatchDeleteRequest(BaseModel):
    ids: List[int]


@router.get("")
def get_favorites(
    chain: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    favorites = list_favorites(db, current_user.id, protocol_type=chain)
    return success_response([FavoriteSchema.model_validate(item) for item in favorites])


@router.post("")
def create_favorite(
    payload: FavoriteCreateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    favorite = add_favorite(
        db,
        current_user.id,
        payload.protocol_name,
        payload.protocol_type,
    )
    return success_response(FavoriteSchema.model_validate(favorite))


@router.delete("/batch")
def remove_favorites_batch(
    payload: FavoriteBatchDeleteRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    deleted, missing = batch_delete_favorites(db, current_user.id, payload.ids)
    return success_response({"deleted": deleted, "missing": missing})


@router.delete("/{favorite_id}")
def remove_favorite(
    favorite_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    delete_favorite(db, current_user.id, favorite_id)
    return success_response({"deleted": True, "id": favorite_id})

File: miniapp/frontend/src/services/favorites.ts
import { apiClient } from './api'

export type Favorite = {
  id: number
  protocolName: string
  protocolType?: string | null
  addedAt: string
}

type RawFavorite = {
  id: number
  protocol_name: string
  protocol_type?: string | null
  added_at: string
}

const mapFavorite = (item: RawFavorite): Favorite => ({
  id: item.id,
  protocolName: item.protocol_name,
  protocolType: item.protocol_type ?? null,
  addedAt: item.added_at,
})

export const getFavorites = async (chain?: string): Promise<Favorite[]> => {
  const response = await apiClient.get<{ data: RawFavorite[] }>('/api/favorites', {
    params: chain ? { chain } : undefined,
  })
  return (response.data.data ?? []).map(mapFavorite)
}

export const addFavorite = async (payload: {
  protocolName: string
  protocolType?: string
}): Promise<Favorite> => {
  const response = await apiClient.post<{ data: RawFavorite }>('/api/favorites', {
    protocol_name: payload.protocolName,
    protocol_type: payload.protocolType,
  })
  return mapFavorite(response.data.data)
}

export const deleteFavorite = async (id: number) => {
  await apiClient.delete(`/api/favorites/${id}`)
  return true
}

export const deleteFavoritesBatch = async (ids: number[]) => {
  await apiClient.delete('/api/favorites/batch', { data: { ids } })
  return true
}
